### PR TITLE
Add check for ASLR when remote VA is enabled

### LINKED ISSUE #604: segv running unit tests on debian 9
I'm running into a segv when running unit tests for SOS 1.3.3 and mvapich2 2.3b on debian 9.3. ```hello``` works with 1 rank but fails with 2. This happens when running the processes locally or remotely to another host. The same behavior occurs when using sockets and psm2 libfabric providers, albeit with a different stack trace.

Any ideas? This works on debian 8. The compiler on debian 9 is gcc 6.3.0 v/s gcc 4.9.2 on debian 8. OSU benchmarks work fine with mvapich2.

```
bsmith@opa7-deb:~$ /usr/shmem/gcc/sandia-openshmem-1.3.3-hfi/bin/oshrun -hosts opa7-deb /usr/shmem/gcc/sandia-openshmem-1.3.3-hfi/share/sandia-openshmem_gcc_hfi/tests/unit/hello
Hello World from 0 of 1
bsmith@opa7-deb:~$ /usr/shmem/gcc/sandia-openshmem-1.3.3-hfi/bin/oshrun -hosts opa7-deb,opa7-deb /usr/shmem/gcc/sandia-openshmem-1.3.3-hfi/share/sandia-openshmem_gcc_hfi/tests/unit/hello
Hello World from 1 of 2
Hello World from 0 of 2

hello:23603 terminated with signal 11 at PC=7ff18dd5cbbc SP=7ffc3cf13930.  Backtrace:
/usr/lib/x86_64-linux-gnu/libfabric/libpsmx2-fi.so(+0x1dbbc)[0x7ff18dd5cbbc]
/usr/lib/x86_64-linux-gnu/libfabric/libpsmx2-fi.so(psmx2_am_atomic_handler_ext+0x28d)[0x7ff18dd644ad]
/usr/lib/x86_64-linux-gnu/libpsm2.so.2(+0x5c5a)[0x7ff18f5afc5a]
/usr/lib/x86_64-linux-gnu/libpsm2.so.2(+0xc48c)[0x7ff18f5b648c]
/usr/lib/x86_64-linux-gnu/libpsm2.so.2(+0x1f48f)[0x7ff18f5c948f]
/usr/lib/x86_64-linux-gnu/libpsm2.so.2(psm2_mq_ipeek+0x98)[0x7ff18f5c49b8]
/usr/lib/x86_64-linux-gnu/libfabric/libpsmx2-fi.so(psmx2_cq_poll_mq+0xb2)[0x7ff18dd4f2d2]
/usr/lib/x86_64-linux-gnu/libfabric/libpsmx2-fi.so(+0x11b7a)[0x7ff18dd50b7a]
/usr/shmem/gcc/sandia-openshmem-1.3.3-hfi/lib/x86_64-linux-gnu/libsma.so.0(+0xf53b)[0x7ff19072e53b]
/usr/shmem/gcc/sandia-openshmem-1.3.3-hfi/lib/x86_64-linux-gnu/libsma.so.0(+0x1111b)[0x7ff19073011b]
/usr/shmem/gcc/sandia-openshmem-1.3.3-hfi/lib/x86_64-linux-gnu/libsma.so.0(+0xd778)[0x7ff19072c778]
/usr/shmem/gcc/sandia-openshmem-1.3.3-hfi/share/sandia-openshmem_gcc_hfi/tests/unit/hello(+0x8bf)[0x557f786a18bf]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf1)[0x7ff18feb02b1]
/usr/shmem/gcc/sandia-openshmem-1.3.3-hfi/share/sandia-openshmem_gcc_hfi/tests/unit/hello(+0x90a)[0x557f786a190a]
[0000] WARN:  init.c:109: shmem_internal_shutdown_atexit
[0000]        shutting down without a call to shmem_finalize
^C[mpiexec@opa7-deb] Sending Ctrl-C to processes as requested
[mpiexec@opa7-deb] Press Ctrl-C again to force abort
```

### PR DESCRIPTION
**Pull request by jdinan**
Sample output:
```
$ mpiexec -np 2 test/unit/hello
[0000] ERROR: ../../src/init.c:231: shmem_internal_init
[0000]        Remote virtual addressing is enabled; however, address space layout randomization
[0000]        is present.  Disable ASLR or rebuild without '--enable-remote-virtual-addressing'.
[0001] ERROR: ../../src/init.c:231: shmem_internal_init
[0001]        Remote virtual addressing is enabled; however, address space layout randomization
[0001]        is present.  Disable ASLR or rebuild without '--enable-remote-virtual-addressing'.

$ mpiexec -np 2 setarch `uname -m` -R test/unit/hello
Hello World from 0 of 2
Hello World from 1 of 2

$ sudo su
# sudo echo "0" > /proc/sys/kernel/randomize_va_space

$ mpiexec -np 2 test/unit/hello
Hello World from 0 of 2
Hello World from 1 of 2

$ mpiexec -np 2 setarch `uname -m` -R test/unit/hello
Hello World from 0 of 2
Hello World from 1 of 2
```

Note that OS X also performs ASLR since 10.7; this does not provide a check for that platform.  We currently disable ASLR on OS X by detecting the platform and setting the `-no_pie` linker flag at configure time.

Closes #604

### CODE REVIEW

**Comment by jdinan** [comment]
Woah, ASLR has been enabled in Travis this whole time.

**Comment by jdinan** [comment]
Also, I goofed and pushed this branch to the main repository.  This is a reminder to clean that up after we merge.

**Comment by jdinan** [comment]
@davidozog This should be ready for review.

The only question I'm still wrangling with is whether this should be a warning or an error.  I ran some debugging tests in my fork to check whether Travis is actually randomizing addresses and we are getting symmetric addresses across PEs even though ASLR is enabled.  This makes me wonder if we should all go buy a lottery tickets or if there is yet another way to disable ASLR that we aren't checking.

**Comment by jdinan** [comment]
I think that the data segment location is not being randomized (perhaps `-no_pie` is default in Ubuntu, but not Debian).  The location where we map the heap is a fixed distance from the data segment, so that is also not random.

**Comment by jdinan** [comment]
Cray tests seem to fail pretty frequently with `icc` in the `SOS_ENABLE_ERROR_TESTS=1 SHMEM_BARRIER_ALGORITHM=auto SHMEM_BCAST_ALGORITHM=auto SHMEM_REDUCE_ALGORITHM=auto SHMEM_COLLECT_ALGORITHM=auto SHMEM_FCOLLECT_ALGORITHM=auto SOS_BUILD_OPTS=\"--enable-error-checking --enable-remote-virtual-addressing --enable-thread-completion\"` config.

Mellanox reduction tests also seem to be consistently timing out in the `SOS_BUILD_OPTS=\"--with-cma --enable-error-checking --enable-profiling --disable-mr-scalable --enable-av-map --enable-remote-virtual-addressing\"` config.

**Comment by jdinan** [comment]
Travis build "passed".  Failure pattern is what I mentioned above for icc.  Shouldn't be necessary to let this run twice (actually, I'm surprised Travis doesn't detect push/pr on the same repository a squash the second testing session).

File: src/init.c [hunk 1]
```diff
@@ -210,6 +214,40 @@ shmem_internal_init(int tl_requested, int *tl_provided)
...
+                                    "is present.  Disable ASLR or rebuild without '--enable-remote-virtual-addressing'.\n"
+                                    RAISE_PE_PREFIX
+                                    "This error message can be disabled by setting SHMEM_DISABLE_ASLR_CHECK or building\n"
+                                    RAISE_PE_PREFIX
+                                    "with --disable-aslr-check.\n",
+                                    shmem_internal_my_pe, shmem_internal_my_pe, shmem_internal_my_pe);
+                }
+            }
+            fclose(aslr);
+        }
```
[the remark this answers is not in the document]
Comment (jdinan) [reply src/init.c]: The check itself is gated by `defined(__linux__)`, so it's strictly for Linux.  I'm not sure it's a warnable condition if this file isn't there.  For instance, this file might not be present if there's a way to build the kernel without ASLR support (I have not checked whether this is possible, just guessing).

File: src/shmem_internal.h [hunk 2]
```diff
@@ -41,6 +41,7 @@ extern void *shmem_internal_data_base;
 extern long shmem_internal_data_length;
 
 #define SHMEM_INTERNAL_HEAP_OVERHEAD (1024*1024)
+#define SHMEM_INTERNAL_DIAG_STRLEN 1024
```
[the remark this answers is not in the document]
Comment (jdinan) [reply src/shmem_internal.h]: I had "diagnostic message" in mind.  Let me know if you want a change here.

File: src/init.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidozog) [reply src/init.c]: Agreed.

File: src/shmem_internal.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidozog) [reply src/shmem_internal.h]: Nah, just wondering.

**Comment by jdinan** [comment]
@davidozog Changes made.  In the course of updating the PR, I noticed that we need to disable PIE on for all Mac builds, not only for remote VA builds.  I thought I tested this pretty carefully, but somehow it didn't show up.  I backed out the relevant changes, putting back `-Wl,-no_pie` for all Mac builds.

**Comment by davidozog** [comment]
Yup, I noticed that and approved it.  Mac should never have pie... ;)  That just means we always disable ASLR on Mac right?

**Comment by jdinan** [comment]
Mellanox tests issue should be fixed now.

**Comment by jdinan** [comment]
@davidozog This is as close to a pass as we're going to get (everything except the icc rows with known issues passed).  Ready to merge when you are.

**Pull request merged**